src/connection.js

```diff
@@ -1274,6 +1306,95 @@ class Connection extends EventEmitter {
     this.transitionTo(this.STATE.SENT_LOGIN7_WITH_STANDARD_LOGIN);
   }
 
+  getRequestOptions() {
+    var tokenHost = '169.254.169.254';
+    var tokenEndpoint = '/metadata/identity/oauth2/token';
+    var clinetIdKeyName = '&client_id=';
+    const apiVersionKeyName = '&api-version=';
+    const clinetId = this.config.authentication.options.MSIclientID;
+    const msiEndpoint = process.env.MSI_ENDPOINT;
+    const msiSecret = process.env.MSI_SECRET;
+    var tokenPath = '';
+    var tokenPort = '';
+    var tokenHeaders = { 'Metadata': true };
+    // Api version supported on Azure VM
+    var apiVersion = '2018-02-01';
+
+    // If MSI_ENDPOINT and MSI_SECRET are defined, used these instead of the hard-coded host
+    if (undefined !== msiEndpoint && undefined !== msiSecret) {
+      // Parse the environment variable as a url and set corresponding request components
+      var appUrl = url.parse(msiEndpoint, true);
+      tokenHost = appUrl.hostname;
+      tokenPath = appUrl.pathname;
+      tokenPort = appUrl.port;
+      tokenHeaders = {
+        'Secret': msiSecret
+      };
+      // Api version supported on Azure App
+      apiVersion = '2017-09-01';
+      clinetIdKeyName = '&clientid=';
+    } else {
+      tokenPath = tokenEndpoint;
+    }
+
+    tokenPath += '?resource=' + this.fedAuthInfoToken.spn;
+
+    // If connection config contians a client id
+    if ('' !== clinetId) {
+      tokenPath += clinetIdKeyName + clinetId;
+    }
+
+    tokenPath += apiVersionKeyName + apiVersion;
+
+    // Append the API version for connection within an Azure VM
+    var options = {
+      host: tokenHost,
+      path: tokenPath,
+      port: tokenPort,
+      method: 'GET',
+      headers: tokenHeaders
+    };
+    return options;
+  }
+
+  getMSIToken() {
+    // Get the token from the azure source
+    http.get(this.getRequestOptions(), (tokenReponse) => {
```

Comment:
What happens if the http request fails (not an endpoint error, but a transport error)? 🤔

You probably need to handle errors by adding an `error` handler to the `ClientRequest` object that is returned by the `http.get` call. See the documentation for `http.get` for an example.